[win][aarch64] Always reserve frame pointers for Arm64 Windows

dpaoliello opened the pull request:
There is no way in Arm64 Windows to indicate that a given function has used the Frame Pointer as a General Purpose Register, as such stack walks will always assume that the frame chain is valid and will follow whatever value has been saved for the Frame Pointer (even if it is pointing to data, etc.).

This change makes the Frame Pointer always reserved when building for Arm64 Windows to avoid this issue.

We will be updating the official Windows ABI documentation to reflect this requirement, and I will provide a link once it's available.

cc @pmsjt

dpaoliello committed:
[win][aarch64] Always reserve frame pointers for Arm64 Windows

efriedma-quic reviewed (commented):
From an ABI perspective, this seems fine: if you want fp to always be valid, sure, lets enforce that.

Do we generate appropriate diagnostics for inline asm?

Does the AArch64 backend actually support this flag?  I briefly tried, and it looks like this doesn't actually do anything.  I think we only implemented support for 32-bit Arm.

Should we make some corresponding change to LLVM so frontends that don't explicitly request some specific frame-pointer behavior follow the ABI by default?

dpaoliello commented:
> Do we generate appropriate diagnostics for inline asm?

Good call, I'll have to look into that.

> Does the AArch64 backend actually support this flag? I briefly tried, and it looks like this doesn't actually do anything. I think we only implemented support for 32-bit Arm.

Ugh, I didn't realize that it didn't. I'll go take care of that.

> Should we make some corresponding change to LLVM so frontends that don't explicitly request some specific frame-pointer behavior follow the ABI by default?

That was going to be my next task.

dpaoliello commented:
I've wired through support for "reserved" and added a hardcoding for Windows to match the existing Darwin check, plus expanding the tests for all cases (and added a negative test, since the existing test didn't use `w29` even when allowed to).

I'll do the check for inline asm in a followup (assuming that the current reservation in LLVM isn't sufficient).

efriedma-quic reviewed (approved):
LGTM

dpaoliello commented:
Failing test is unrelated: <https://github.com/llvm/llvm-project/issues/146781>

DavidSpickett commented:
I've reverted this due to a test failure on our Linux bots: https://lab.llvm.org/buildbot/#/builders/17/builds/9292

And if I understand correctly, flang needs another patch before it can build properly on Windows:
```
FAILED: include/flang/__fortran_builtins.mod C:/Users/tcwg/llvm-worker/flang-arm64-windows-msvc/build/include/flang/__fortran_builtins.mod 
C:\WINDOWS\system32\cmd.exe /C "cd /D C:\Users\tcwg\llvm-worker\flang-arm64-windows-msvc\build\tools\flang\tools\f18 && C:\Users\tcwg\scoop\apps\cmake\3.29.2\bin\cmake.exe -E make_directory C:/Users/tcwg/llvm-worker/flang-arm64-windows-msvc/build/include/flang && C:\Users\tcwg\llvm-worker\flang-arm64-windows-msvc\build\bin\flang.exe -cpp -fsyntax-only -module-dir C:/Users/tcwg/llvm-worker/flang-arm64-windows-msvc/build/include/flang C:/Users/tcwg/llvm-worker/flang-arm64-windows-msvc/llvm-project/flang/module/__fortran_builtins.f90"
error: invalid value 'reserved' in '-mframe-pointer=reserved'
```
As covered by https://github.com/llvm/llvm-project/issues/146802.

DavidSpickett commented:
The actual error is the same as seen elsewhere:
```
$ /home/david.spickett/build-llvm-aarch64/bin/flang --target=aarch64-pc-windows-msvc -Werror ../llvm-project/flang/test/Semantics/windows.f90
flang-21: warning: unable to find a Visual Studio installation; try running Clang from a developer command prompt [-Wmsvc-not-found]
error: invalid value 'reserved' in '-mframe-pointer=reserved'
```
It was obscured by this test running the compiler via a Python script.

dpaoliello commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -518,6 +518,27 @@ bool AArch64FrameLowering::hasFPImpl(const MachineFunction &MF) const {
+
+  // These OSes require the frame chain is valid, even if the current frame does
+  // not use a frame pointer.
+  if (TT.isOSDarwin() || TT.isOSWindows())
```
I'm not familiar with how flang works, but if it's going to scrape(?) the command line args from clang, then it needs to be able to handle all the same command line arguments that clang does.

This is not a bug in clang: this is a bug in flang and would be hit if anyone tried building Arm32 code in flang.
